### ISSUE TITLE
Does not work on node 17 or above, requires node 16

### ISSUE THREAD

**Issue by Charuru** [badeball/karma-jsdom-launcher#39]
Seems like this is known? https://github.com/badeball/karma-jsdom-launcher/commit/f78b1b55f9e9dc5437e28bcf7db52c6b77946c46

Is there any plan to fix this or is it just stuck on 16 forever?

**Comment by badeball** [comment]
> Seems like this is known? https://github.com/badeball/karma-jsdom-launcher/commit/f78b1b55f9e9dc5437e28bcf7db52c6b77946c46

No, it just means that 16 was most recent at the time of writing. Is there something not working on 17+?

**Comment by Charuru** [comment]
Oh okay, 17&18 was out at that time so I assumed. Just doesn't start for me on 16+ and there are other reports on google.

https://stackoverflow.com/questions/70278340/unable-to-run-karmajsdom-tests-on-nodejs-17-x

```
26 10 2022 11:46:04.352:WARN [karma]: No captured browser, open http://localhost:9876/
26 10 2022 11:46:04.378:INFO [karma-server]: Karma v6.4.1 server started at http://localhost:9876/
26 10 2022 11:46:04.379:INFO [launcher]: Launching browsers jsdom with concurrency unlimited
26 10 2022 11:46:04.381:INFO [launcher]: Starting browser jsdom
26 10 2022 11:46:04.401:ERROR [karma-server]: UnhandledRejection: Error: connect ECONNREFUSED ::1:9876
    at TCPConnectWrap.afterConnect [as oncomplete] (node:net:1195:16)
26 10 2022 11:46:04.401:ERROR [karma-server]: Error: connect ECONNREFUSED ::1:9876
    at TCPConnectWrap.afterConnect [as oncomplete] (node:net:1195:16) {
  errno: -61,
  code: 'ECONNREFUSED',
  syscall: 'connect',
  address: '::1',
  port: 9876
}
```

Also what do you think about converting this to happy-dom instead?

@badeball from what I see then callback never happens.

```
      jsdom.JSDOM.fromURL(url, jsdomOptions).then(function (dom) {
        self.window = dom.window;
      });

```

**Comment by badeball** [comment]
[This](https://github.com/nodejs/node/issues/40537) might be of interest to you.

Can you try running tests with `NODE_OPTIONS=--dns-result-order=ipv4first ..`?

> Also what do you think about converting this to happy-dom instead?

I think a happy-dom integration should exist as a separate package.

**Comment by Charuru** [comment]
That does fix it, are you able to use 16+ without that option?

**Comment by badeball** [comment]
It's a bit up to you. You can either run your app and have it listen on both ipv4 and ipv6, in which case all node versions are going to work or you have to tell node and jsdom what you mean when you ask it to access "localhost". You can change the default from the runtime like so:

```js
import dns from "node:dns";
dns.setDefaultResultOrder("ipv4first");
```

**Comment by badeball** [comment]
Wait, karma _is_ the app in this context (I forget what repo I am in..). Maybe we can make some assumptions about it. Leaving this open for now.

**Comment by Charuru** [comment]
I don't understand, where do I do this? Shouldn't it be handled by the launcher?

Other launchers like ChromeCanaryHeadless that I'm also using works fine without any changes.
